benches/end_to_end.rs

```diff
@@ -1,385 +1,472 @@
+            });
+        } else {
+            b.iter(|| {
+                let futs = (0..self.parallel_cnt).map(|_| {
+                    let req = make_request(&self, &rt, &addr);
+                    send_request(req)
+                });
+
+                rt.block_on(join_all(futs));
+            });
+        }
+    }
+}
+
+//
+// Creates a request, for being sent via the request_sender
+//
+fn make_request(opts: &Opts, rt: &tokio::runtime::Runtime, addr: &SocketAddr) -> Request<Body> {
+    let url: hyper::Uri = format!("http://{}/hello", addr).parse().unwrap();
+    let chunk_cnt = opts.request_chunks;
+    let body = if chunk_cnt > 0 {
+        let (mut tx, body) = Body::channel();
+        let chunk = opts
+            .request_body
+            .expect("request_chunks means request_body");
+        rt.spawn(async move {
+            for _ in 0..chunk_cnt {
+                tx.send_data(chunk.into()).await.expect("send_data");
+            }
+        });
+        body
+    } else {
+        opts.request_body
+            .map(Body::from)
+            .unwrap_or_else(Body::empty)
+    };
+    let mut req = Request::new(body);
+    *req.method_mut() = opts.request_method.clone();
+    *req.uri_mut() = url.clone();
+    req
+}
+
+//
+// Prepares a client (request_sender) for sending requests to the given addr
+//
+fn prepare_client(
+    rt: &tokio::runtime::Runtime,
+    opts: &Opts,
+    addr: &SocketAddr,
+) -> SendRequest<crate::Body> {
+    let mut builder = Builder::new();
+    builder
+        .http2_only(opts.http2)
+        .http2_initial_stream_window_size(opts.http2_stream_window)
+        .http2_initial_connection_window_size(opts.http2_conn_window)
+        .http2_adaptive_window(opts.http2_adaptive_window);
+
+    let (request_sender, connection) = rt.block_on(async move {
+        let target_stream = TcpStream::connect(addr).await.unwrap();
+
+        builder.handshake(target_stream).await.unwrap()
+    });
+    rt.spawn(async move {
+        if let Err(e) = connection.await {
+            panic!("Error in connection: {}", e);
+        }
+    });
+    request_sender
+}
+
+//
+// Spawns a server in the background
+//
+fn spawn_server(rt: &tokio::runtime::Runtime, opts: &Opts) -> SocketAddr {
+    let addr: SocketAddr = "127.0.0.1:0".parse().unwrap();
+    let tcp_listener = rt.block_on(async move { TcpListener::bind(addr).await.unwrap() });
+    let local_addr = tcp_listener.local_addr().unwrap();
+    let mut http = Http::new();
+    //configure http based on opts
+    http.http2_only(opts.http2)
+        .http2_initial_stream_window_size(opts.http2_stream_window)
+        .http2_initial_connection_window_size(opts.http2_conn_window)
+        .http2_adaptive_window(opts.http2_adaptive_window);
+
+    let (http2, http2_stream_window, http2_conn_window, http2_adaptive_window) = (
+        opts.http2,
+        opts.http2_stream_window,
+        opts.http2_conn_window,
+        opts.http2_adaptive_window,
+    );
```

Comment:
If you `#[derive(Clone)]` for the `Opts` struct you can clone opts here and then use it directly without destructuring. (Except for the `opts.response_body`, which has to be destructured before being passed to the `service_fn` closure.)

```rust
 let opts = opts.clone();
 let body = opts.response_body;
```